hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WindowSizeGCDCalculator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt;
+
+import com.hazelcast.jet.impl.util.Util;
+import com.hazelcast.jet.sql.impl.opt.physical.FullScanPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.PhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.ShouldNotExecuteRel;
+import com.hazelcast.jet.sql.impl.opt.physical.SlidingWindowAggregatePhysicalRel;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import org.apache.calcite.rel.BiRel;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelVisitor;
+import org.apache.calcite.rel.core.Union;
+
+import javax.annotation.Nullable;
+
+public class WindowSizeGCDCalculator {
+    static final long DEFAULT_THROTTLING_FRAME_SIZE = 10L;
+
+    private final GCDCalculatorVisitor visitor;
+    private final SlidingWindowDetector swDetector;
+
+    public WindowSizeGCDCalculator(ExpressionEvalContext eec) {
+        this.visitor = new GCDCalculatorVisitor(eec);
+        this.swDetector = new SlidingWindowDetector();
+    }
+
+    public void calculate(PhysicalRel rel) {
+        if (shouldRun(rel)) {
+            visitor.go(rel);
+        }
+    }
+
+    public long get() {
+        return visitor.gcd > 0 ? visitor.gcd : DEFAULT_THROTTLING_FRAME_SIZE;
```

Comment:
We should max the GCD at 100ms for large window sizes. This is good for observing the progression of event time in metrics. 10 WMs per second are negligible.